File: [K]graph/[K]graph-bfs/133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
from collections import deque
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return node
        old_new = {}
        queue = deque([])
        visited = set()
        queue.append(node)
        visited.add(node)
        while queue:
            old_node = queue.popleft()
            if old_node not in old_new:
                old_new[old_node] = Node(old_node.val)
            for neighbor in old_node.neighbors:
                if neighbor not in old_new:
                    old_new[neighbor] = Node(neighbor.val)
                old_new[old_node].neighbors.append(old_new[neighbor])
                if neighbor not in visited:
                    queue.append(neighbor)
                    visited.add(neighbor)
        return old_new[node]
    
# time O(n)
# space O(n)
# using graph and bfs with single source and hashmap

File: [K]graph/[K]graph-bfs/133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        HashMap<Node, Node> oldToNew = new HashMap<>();
        if (node == null) {
            return node;
        }
        oldToNew.put(node, new Node(node.val));
        ArrayDeque<Node> queue = new ArrayDeque<>();
        queue.offer(node);
        while (!queue.isEmpty()) {
            Node oldNode = queue.poll();
            Node newNode = oldToNew.get(oldNode);
            for (Node neighbor: oldNode.neighbors) {
                if (!oldToNew.containsKey(neighbor)) {
                    oldToNew.put(neighbor, new Node(neighbor.val));
                    queue.offer(neighbor);
                }
                Node newNeighbor = oldToNew.get(neighbor);
                newNode.neighbors.add(newNeighbor);
            }
        }
        return oldToNew.get(node);
    }
}

// time O(n)
// space O(n)
// using graph and bfs with single source and hashmap